### PR TITLE
OO Mailroom assignment refactored Josh Hoff

### PR DESCRIPTION
**Pull request by josh1209**
@natasha-aleksandrova Please take a look and let me know what you think. I refactored my OO mailroom assignment to better utilize OO concepts.

Thank you,

-Josh

### CODE REVIEW

File: students/Josh_HOff/lesson09/mailroom_lesson_09.py [hunk 1]
```diff
@@ -4,70 +4,149 @@
...
+        sys.exit()
+        
+    def continuing(self):
+        print('Try Again.\n')
+
+#this class handles all donor management
+class Donor(object):
+    def __init__(self=None, name = ''):
+        global donors
+        self._donors = donors
```
Comment (natasha-aleksandrova) [inline students/Josh_HOff/lesson09/mailroom_lesson_09.py]: this class is for a single donor object; what is the purpose of donors attribute?

File: students/Josh_HOff/lesson09/mailroom_lesson_09.py [hunk 2]
```diff
@@ -4,70 +4,149 @@
...
+        self._gifts = len(self._donors.get(name))
+        self._total = sum(self._donors.get(name))
+        self._average = round((self._total / self._gifts), 2)
+        self._recent_gift = self._donors.get(name)[-1]
+        self._first_gift = self._donors.get(name)[0]
+                
+    @property
+    def gifts(self):
+        return self._gifts
```
Comment (natasha-aleksandrova) [inline students/Josh_HOff/lesson09/mailroom_lesson_09.py]: no need for a property - you're doing anything special to it? access it directly as an attribute

File: students/Josh_HOff/lesson09/mailroom_lesson_09.py [hunk 3]
```diff
@@ -4,70 +4,149 @@
...
+        
+    @gifts.deleter
+    def gifts(self):
+        del self._gifts
```
Comment (natasha-aleksandrova) [inline students/Josh_HOff/lesson09/mailroom_lesson_09.py]: what is the purpose of the deleter?

File: students/Josh_HOff/lesson09/mailroom_lesson_09.py [hunk 4]
```diff
@@ -4,70 +4,149 @@
...
+    def recent_gift(self):
+        del _recent_gift
+        
+    @property
+    def first_gift(self):
+        return self._first_gift
+        
+    @first_gift.deleter
+    def first_gift(self):
+        del self._first_gift
```
Comment (natasha-aleksandrova) [inline students/Josh_HOff/lesson09/mailroom_lesson_09.py]: same comment for all of these properties

File: students/Josh_HOff/lesson09/mailroom_lesson_09.py [hunk 5]
```diff
@@ -4,70 +4,149 @@
...
+        
+        
+class DonorCollection(object):
+    def __init__(self=None):
+        pass
```
Comment (natasha-aleksandrova) [inline students/Josh_HOff/lesson09/mailroom_lesson_09.py]: this isn't really object oriented, you're not holding any state? 

staticmethods are just helper functions that relate to a class, not necessary part of an object.

File: students/Josh_HOff/lesson09/mailroom_lesson_09.py [hunk 6]
```diff
@@ -4,70 +4,149 @@
...
+    @staticmethod
+    def letters():
+        tab = '    '
+        global donors
+        for name, val in donors.items():
+            with open(f'{name}.txt', 'w') as outfile:
+                d = Donor(name)
+                donation = d.total_donations
+                val = d.recent_gift
+                outfile.write(f'Dear {name}, \n\n{tab}Thank you very much for your most recent donation \
```
Comment (natasha-aleksandrova) [inline students/Josh_HOff/lesson09/mailroom_lesson_09.py]: the class usage here (`Donor` class) does not really provide you any benefit

File: students/Josh_HOff/lesson09/mailroom_lesson_09.py (hunk 4 above)
Comment (josh1209) [reply students/Josh_HOff/lesson09/mailroom_lesson_09.py]: @natasha-aleksandrova I'll remove the deleters

**Comment by josh1209** [comment]
I thought the point was to make donor attributes retrievable like in the circle class, so in ipython I could feed the terminal c = Donor(‘Josh Hoff’) and be able to pull up various stats such as average of gifts, total dollar amount of gifts, most recent gift received, etc. using commands like c.average, c.total_donations, c.first_gift.

I’ll look at some more examples. I’m very clearly misunderstanding the idea behind OO programming.

Thank you,

-Josh

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows 10

________________________________

**Pull request merged**